### PR TITLE
Add base configuration for ActiveRecord, untested yet

### PR DESCRIPTION
**Pull request by terry90**
Adds ActiveRecord support.

Some things are missing still:

We should add this middleware to ensure the connection management is well done
```ruby
module ActiveRecord
  module Rack
    # ActiveRecord >= 5 removes the Pool management
    class ConnectionManagement
      def initialize(app)
        @app = app
      end

      def call(env)
        response = @app.call(env)
        response[2] = ::Rack::BodyProxy.new(response[2]) do
          ActiveRecord::Base.clear_active_connections!
        end

        return response
      rescue Exception
        ActiveRecord::Base.clear_active_connections!
        raise
      end
    end
  end
end
```

Maybe we could add a `common.rb` for databases.yml of ORMs, just overriding variables like `@dbname`, `@username` etc.. to generates the same config file with custom fields.

### CODE REVIEW

**Comment by LeFnord** [comment]
@terry90 … first: many thanks for your work

the middleware you mentioned above is only for AR, right?
when it would be needed, why you don't add it?
I'm not relay the AR guru, sorry, so it could be that I misunderstand it …

to your second point, I totally agree with, and I'll have a closer look on it, after merging this (your) PR,
so the requirements are more apparent, and it is easier to write specs 😉

File: lib/starter/version.rb [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # frozen_string_literal: true
 
 module Starter
-  VERSION = '0.7.0-sequel'
+  VERSION = '0.7.0-activerecord'
```
Comment (LeFnord) [inline lib/starter/version.rb]: please change to: `0.7.0`, will cut a release after this PR was merged

**Comment by terry90** [comment]
I'm wondering if I should add the middleware directly in the new! method of the builder, like this:

```ruby
def new!(name, source, destination, options = {})
  @prefix = options[:p] || 'api'
  @resource = name
  @destination = destination

  FileUtils.copy_entry source, destination

  replace_static(File.join('script', 'server'), "API-#{resource}")
  replace_static(File.join('api', 'base.rb'), ":#{prefix}")
  replace_static(File.join('spec', 'requests', 'root_spec.rb'), prefix)
  replace_static(File.join('spec', 'requests', 'documentation_spec.rb'), prefix)

  # Here, replace_static of config.ru either replace with blank line when not active record
  # or replace with use ActiveRecord::ConnectionManagement if orm is ActiveRecord
  # + Add connection_management.rb in initializers

  Orms.build(destination, options[:orm]) if options[:orm]
  self
end
```

What do you think about this ? Is it the preferred way ?
It will certainly lead to bulk code blocks if we do not dry up here

**Comment by LeFnord** [comment]
is it needed in `config.ru`? think placing it into `initializers` would be fine … 
> `# Here, replace_static of config.ru either replace with blank line when not ActiveRecord`
> ` # or replace with use ActiveRecord::Rack::ConnectionManagement if orm is ActiveRecord`
> ` # + Add connection_management.rb in initializers (the code I posted above in the PR description)`

not replacing, better [`append_to_file `](https://github.com/LeFnord/grape-starter/blob/master/lib/starter/builder/orms.rb#L34-L37) … 

and yeah, have refactor all the file stuff into `FileFoo` 😉

**Comment by terry90** [comment]
We can't use `append_to_file`, I think we  have to provide use Custom::Middleware before  running the application `run App.new`

**Comment by LeFnord** [comment]
mmh, ok … this means, an initializer must be created, and this entry in `config.ru`, right?

**Comment by LeFnord** [comment]
@terry90 … please make it ready, so I can mörge it, I trust you, that you are doing the right stuff 😉
if it is needed, then it should be added

**Comment by terry90** [comment]
Done, entry in  config.ru OK. I added a add_middleware function for ORMs. Maybe I should put it somewhere else. Anyway, I think lib_klass_name is not working when adding a new resource.
IS it working  for you ? `class Model` should be  `class Model < Sequel / ActiveRecord::Base`

**Comment by LeFnord** [comment]
thanks @terry90 … I'll have a look on the `lib_klass_name` stuff

**Comment by LeFnord** [comment]
please rebase against master …
for the `lib_klass_name` stuff, there is a [spec](https://github.com/LeFnord/grape-starter/blob/master/spec/lib/starter/names_spec.rb#L33-L41) … mmh

File: lib/starter/builder/templates/activerecord.rb [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  module Templates
+    module ActiveRecord
+      def model_klass
+        'ActiveRecord::Base'
+      end
+
+      def initializer
+        <<-FILE.strip_heredoc
+        require 'yaml'
+        require 'erb'
```
Comment (LeFnord) [inline lib/starter/builder/templates/activerecord.rb]: please add: `require 'active_record'`

File: lib/starter/builder/templates/activerecord.rb [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+        config_dir = File.expand_path('../', __FILE__)
+        config_content = File.read(File.join(config_dir, 'database.yml'))
```
Comment (LeFnord) [inline lib/starter/builder/templates/activerecord.rb]: change to: `config_content = File.read(File.join('config', 'database.yml'))` and remove Line#15

File: lib/starter/builder/templates/activerecord.rb [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+        db_conf = YAML.safe_load(ERB.new(config_content).result)
+        env = ENV['RACK_ENV'] || 'development'
+
+        ActiveRecord::Base.establish_connection db_conf[env]
+        logger = if %w[development test].include? env
+          Logger.new(File.open('log/db.log', 'a'))
```
Comment (LeFnord) [inline lib/starter/builder/templates/activerecord.rb]: replace with:
```ruby
log_dir = File.join(Dir.getwd, 'log')
log_file = File.join(log_dir, 'db.log')
FileUtils.mkdir(log_dir) unless Dir.exist?(log_dir)
FileUtils.touch(log_file) unless File.exist?(log_file)
Logger.new(File.open(log_file, 'a'))
```

File: lib/starter/builder/templates/activerecord.rb [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+          end
+        end
+        FILE
+      end
+
+      def config # TODO: Dry (Same config for Sequel)
+        <<-FILE.strip_heredoc
+        # ActiveRecord Database Configuration
+        development:
+          adapter: 'sqlite'
```
Comment (LeFnord) [inline lib/starter/builder/templates/activerecord.rb]: change to: `sqlite3`

File: lib/starter/builder/templates/activerecord.rb [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+          host: localhost
+          port: 27017
+          database: "db/development.sqlite3"
+          username:
+          password:
+
+        test:
+          adapter: 'sqlite'
```
Comment (LeFnord) [inline lib/starter/builder/templates/activerecord.rb]: change to: `sqlite3`

File: lib/starter/builder/templates/activerecord.rb [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+          host: localhost
+          port: 27017
+          database: "db/test.sqlite3"
+          username:
+          password:
+
+        production:
+          adapter: 'sqlite'
```
Comment (LeFnord) [inline lib/starter/builder/templates/activerecord.rb]: change to: `sqlite3`

File: lib/starter/builder/templates/activerecord.rb (hunk 2 above)
Comment (terry90) [inline lib/starter/builder/templates/activerecord.rb]: Ain't it automatically required by bundler ?

This is the order:

```
Bundler.require :default, ENV['RACK_ENV']

Dir[File.expand_path('../../config/initializers/*.rb', __FILE__)].each do |initializer|
  require initializer
end
```

File: lib/starter/builder/templates/activerecord.rb (hunk 3 above)
Comment (terry90) [inline lib/starter/builder/templates/activerecord.rb]: done

File: lib/starter/builder/templates/activerecord.rb (hunk 2 above)
Comment (LeFnord) [inline lib/starter/builder/templates/activerecord.rb]: yeah, that was also my though, but no … ☹️

we can do the fine-tuning, after it is merged, ok?

File: lib/starter/builder/templates/activerecord.rb (hunk 4 above)
Comment (terry90) [inline lib/starter/builder/templates/activerecord.rb]: Done, but I got rid of the touch file:
Thanks to:

```plaintext
"a"  Write-only, each write call appends data at end of file.
     Creates a new file for writing if file does not exist.
```

File: lib/starter/builder/templates/activerecord.rb (hunk 2 above)
Comment (terry90) [inline lib/starter/builder/templates/activerecord.rb]: Waiting for your feedback before doing this

Comment (terry90) [inline lib/starter/builder/templates/activerecord.rb]: Alright then :+1:

Comment (terry90) [inline lib/starter/builder/templates/activerecord.rb]: Done

**Pull request merged**